core/src/core/verifier_cache.rs

```diff
@@ -15,52 +15,90 @@
 //! VerifierCache trait for batch verifying outputs and kernels.
 //! We pass a "caching verifier" into the block validation processing with this.
 
-use lru_cache::LruCache;
+use cuckoocache::cache::{Cache, SaltedHasher};
 
-use crate::core::hash::{Hash, Hashed};
+use crate::core::hash::{DefaultHashable, Hash, Hashed, ZERO_HASH};
 use crate::core::{Output, TxKernel};
+use crate::util::RwLock;
+use rand::prelude::random;
 
+struct VerifierHasher {
+	salt: Hash,
+}
+impl<E: DefaultHashable> SaltedHasher<E> for VerifierHasher {
+	fn new() -> Self {
+		Self {
+			salt: Hash::from_vec(&random::<[u8; 32]>()[..]),
+		}
+	}
+	fn hashes(&self, e: &E) -> [u32; 8] {
+		let mut hs = [0u32; 8];
+
+		for (v, h) in (self.salt, e)
```

Comment:
What's going on with this funky hasher? Can't we just use our own `hash` impl?

Reply:
We do use our own hash implementation -- `(self.salt, e).hash()`. The concern is that the hashes need to be salted with a private salt. Otherwise I can easily DoS the network by finding a hash cycle in the cache. The rest of the funkyness is just to cast from [u8;32] -> [u32;8]. I don't know of any real way to do that? Maybe we could do...

```rust
let mut x = [0u32; 8];
{
  let y : &mut [u8; 32] = unsafe{std::mem::transmute(&mut x)}
  y = (self.salt, e).hash();
}
x
```

but then we're a smidgen unsafe.

Reply:
Ah - I missed the actual `hash()` call. :+1:

I get what you're saying - this way everyone hashes differently.
Let me think on this a bit. Makes we wonder if we need to consider doing this elsewhere.

Reply:
Actually - can we use the `siphash` stuff in `id.rs`? This would give us a way of getting short hashes, salted with a private value locally.

Reply:
And related - if the DoS is an issue, do we need to be doing something similar to this for the db/index itself as we store various things based on their hashes in the db?

Reply:
agree with antiochp that using a keyed hash function like our existing siphash24 is cleaner than this salted construct.

Reply:
In this case we need 256 bits of hash, so we'd have to siphash with a lot of different keys.

I suppose we could get away with using fewer bits, but we need at least enough to support 8x the number of bits needed to address our table.

Most likely we should salt in the DB too.

Reply:
In this case we need 256 bits of hash, so we'd have to siphash with a lot of different keys.

I suppose we could get away with using fewer bits, but we need at least enough to support 8x the number of bits needed to address our table.

Most likely we should salt in the DB too.

Reply:
in what circumstance would 128 bits of siphash key not suffice?

note that our Cuckoo Cycle proof of work uses a variant of siphash24 with a full 256-bit key,
so we could also consider using that instead of the standard siphash24...

Reply:
The way that cuckoocache works we use 8 insertion indexes. Our method of generating them is to get a 256 bit hash. And split it up into 8 and then use lemire's hashing trick to project each of 8 into the range of the table. Because we want to support up to a 2**32 element table, we need to have 32 bits per index. 

We also store only the hashed value, not the value itself, so it's important that the hash is collision resistant.

Reply:
The way that cuckoocache works we use 8 insertion indexes. Our method of generating them is to get a 256 bit hash. And split it up into 8 and then use lemire's hashing trick to project each of 8 into the range of the table. Because we want to support up to a 2**32 element table, we need to have 32 bits per index. 

We also store only the hashed value, not the value itself, so it's important that the hash is collision resistant.

Reply:
The way that cuckoocache works we use 8 insertion indexes. Our method of generating them is to get a 256 bit hash. And split it up into 8 and then use lemire's hashing trick to project each of 8 into the range of the table. Because we want to support up to a 2**32 element table, we need to have 32 bits per index. 

We also store only the hashed value, not the value itself, so it's important that the hash is collision resistant.

Reply:
The way that cuckoocache works we use 8 insertion indexes. Our method of generating them is to get a 256 bit hash. And split it up into 8 and then use lemire's hashing trick to project each of 8 into the range of the table. Because we want to support up to a 2**32 element table, we need to have 32 bits per index. 

We also store only the hashed value, not the value itself, so it's important that the hash is collision resistant.

Reply:
The blake2b crate we use may also allow us to get 256 bits in a single digest?

Reply:
We do already, I'm not computing the hash multiple times, just once?

Reply:
considering the difficulty of getting more than 64 approved bits of siphash output, and the need for 8x32 bits for cuckoocache, I think using blake2b as a keyed hash function may be best?!

Reply:
I'm not using siphash, I'm using blake2b as a salted hash. I'm fine using keyed instead of salted, we just don't have support for it in the code currently. The hash method uses the Hashed trait. At least I'm fairly certain I am -- is there a reason you think otherwise?

@antiochp can I add one point to the 'we should rename the hash method name' tally?

Reply:
Checking the blake2b documentation at https://blake2.net/blake2.pdf, I see that it supports both keyed hashing, and a salt. The difference is that the salt is xored into the IV (initialization vector), while the key is processed as the first message block. The former seems preferable as long as one is fine with 16 bytes salts. So Jeremy was ahead of me all along:-) Sorry for the unnecessary siphash discussion.

I would just prefer the salt to be passed into the blake2b constructor as part of the blake2b parameter block, rather than being passed around as the first element of a rust pair to be hashed.

Reply:
Fair enough -- it just requires some additional code to support using keyed hashing, which we don't currently have (but the blake2 crate does).

I agree it's cleaner to pass it into the parameter block.